src/sentry/integrations/slack/client.py

```diff
@@ -46,7 +47,20 @@ def track_response_data(self, code, span, error=None, resp=None):
     def request(self, method, path, headers=None, data=None, params=None, json=False, timeout=None):
         # TODO(meredith): Slack actually supports json now for the chat.postMessage so we
         # can update that so we don't have to pass json=False here
-        response = self._request(method, path, headers=headers, data=data, params=params, json=json)
-        if not response.json.get("ok"):
-            raise ApiError(response.get("error", ""))
+        while True:
+            response = self._request(
+                method, path, headers=headers, data=data, params=params, json=json
+            )
+            if response.json.get("ok"):
+                break
+
+            if not response.json.get("ok"):
+                # Check if we were rate limited
+                retry_after = response.headers.get("retry-after")
+                if retry_after:
+                    self.logger.info("rule.slack.rate_limited", extra={"retry_after": retry_after})
+                    time.sleep(float(retry_after))
```

Comment:
I don't see where the actual Celery task lies in that executes this function but if this is running in a worker, instead of having an infinite loop with sleep, we can should just retry the task: https://docs.celeryproject.org/en/stable/userguide/tasks.html#retrying

Reply:
For message sending that makes sense to me, but for paginated API's I don't think you would want to start the listing over again. I guess it doesn't make a lot of sense to implement this in the client. If it were to be anywhere, maybe this _only_ applies to the channel listing?

Reply:
@beeftornado - oh right, I forgot the actual problem you were hitting. Channel listing should not be happening in a worker thread tho. I'd say for that we return the `retry_after` value to the client and retry from the client side to not consume server resources. What do you think?

Reply:
Sorry, I think this was my fault for the ambiguity. When I was talking about the client, I was actually referring to the `SlackClient` class. 

I assume when you're talking about `client`, you are referring to the front-end? I don't think it would be possible to retry from there, right, because the front-end doesn't directly resolve the channel name to channel id; the backend does it as part of the rule update.

https://github.com/getsentry/sentry/blob/6adfd666d0ab9d768bcf29b1dae1a43eb4da077a/src/sentry/integrations/slack/utils.py#L403-L413

If during the channel name lookup, at page 3/5, a request fails for the next page, I would expect not to have to throw away the previously fetched pages, or at least remember that the channel name didn't appear on those pages.

It sounds like the only risk-free way of dealing with the rate limiting is to maintain the state the of the lookup somewhere, if a 429 is encountered then it schedules a task in the future to resume where the lookup left off. Am I following the conversation correctly?

Reply:
Ugh, this is really tricky to figure out. Slack not having a proper search API is ridiculous. I guess we should just add `exclude_archived=True` to begin with. I also realized this method has zero caching and has an `async_lookup` parameter which just increases the timeout. 

I think if we just remove archived channels and add some app-wide long-term caching, we'd go a long way to improve the situation. What do you think?

Reply:
> I also realized this method has zero caching and has an async_lookup parameter which just increases the timeout

Heh :)

> remove archived channels and add some app-wide long-term caching

I think this would be sufficient for like 99.9% of cases, including ours. Excluding archived would probably be enough for our use case, but I wasn't sure what Sentry's behavior would be if it tried to lookup a channel for an existing rule, and that channel gets archived. Would it break all alerts for the project or just the one rule? The guess/test/revise process is a little tedious in my current setup. 

Are we agreed that this pull should be closed?

Reply:
Agree that this pull should be closed.

@mattrobenolt do you think if we started caching Slack channel ID lookups, would that cause a big swell on our caches?

@markstory @wedamija - If we start ignoring archived channels, would that cause issues during alerting time if the alert refers to an archived channel? IMO we shouldn't even be looking up channel ids during alerting as the ids never change, right?

Reply:
@BYK `exclude_archived` doesn't actually help all that much actually, assuming you're talking about paginating through channels. All this does is not return the archived channels on that particular page, but it doesn't change the number of pages to be returned at all. I guess this does help in terms of having less rows to look at on our side, but I'm not sure it makes a major improvement to the slowness of this lookup.

As an example:
Page 2 has 1000 results, and 995 of them are archived. If we use `exclude_archived=True`, then page 2 will return just 5 results, and we'll still have to continue on with pagination on page 3. At least, this was how it worked when I last tested it, which was admittedly a while ago.

I'm not sure what we want to do here in terms of caching - just cache each page for some short period of time so that on retries things will be a little faster? I think if we cache for say 10-20 minutes that would be fine. We don't do these channel lookups often, so I don't see there being any real issue of taking up too much cache.

> If we start ignoring archived channels, would that cause issues during alerting time if the alert refers to an archived channel? IMO we shouldn't even be looking up channel ids during alerting as the ids never change, right? 
Yeah this will be fine, the lookup only happens when creating a rule so that we can get the internal id. On firing we can just look up the internal id directly. If that channel has been archived since the rule was created we'll fail to send the message, but that's expected.

Reply:
@wedamija the `exclude_archived` behavior you explained is very bizarre to me. I thought it would filter archived channels first and then paginate but what you describe feels like Slack 😁 

Re caching, my point was once you resolve a channel name to an ID, that never changes so why look up again anyway? We can even put them in persistent storage (or something semi-persistent like Redis?)

Reply:
> Re caching, my point was once you resolve a channel name to an ID, that never changes so why look up again anyway? We can even put them in persistent storage (or something semi-persistent like Redis?)

I mean in theory that is probably true 99% of the time. But channels can be renamed and I assume they keep the same id, which would end up causing problems if we cached these long term. I'd be open to caching for a longer period of time though.

Reply:
Yep they do keep the same ID if renamed. Caching would help with reducing the number of lookups we do so +1 on that. And maybe that's enough -- who knows. If it's not (or maybe in addition), when `get_channel_id_with_timeout` does throw a 429, we could catch that and retry whenever Slack's API tells us to retry cause we already put it in a queue if it does timeout.

But either way, I do agree that this pull should be closed. Sounds like we need to do some experimentation + monitoring.